src/android/Dgram.java

```diff
@@ -75,198 +48,194 @@ public void run() {
-                    } catch (Exception innerException) {
-                        Log.e(TAG, "Receive exception:" + innerException.toString());
-                    } finally {
-                        return;
+                    if (onMessageCallback != null) {
+                        sendMessageResult(message, address, port);
                     }
+                } catch (Exception e) {
+                    Log.d(TAG, "Received exception:" + e.toString());
+                    sendMessageErrorResult(e);
                 }
             }
         }
     }
 
+    private void sendMessageResult(String message, String address, int port) {
+        JSONObject payload = new JSONObject();
+
+        try {
+            payload.put("message", message.trim());
+            payload.put("address", address);
+            payload.put("port", port);
+            CallbackUtil.sendPluginResult(onMessageCallback, new PluginResult(PluginResult.Status.OK, payload));
+        } catch (Exception e) {
+            PluginResult result = new PluginResult(
+                    PluginResult.Status.ERROR,
+                    "Error occurred while creating onMessage payload: " + e.toString()
+            );
+            CallbackUtil.sendPluginResult(onMessageCallback, result);
+        }
+    }
+
+    private void sendMessageErrorResult(Exception e) {
+        try {
+            JSONObject payload = new JSONObject();
+
+            payload.put("error", e.toString());
+
+            CallbackUtil.sendPluginResult(onMessageCallback,
+                    new PluginResult(PluginResult.Status.ERROR, payload)
+            );
+        } catch (Exception exception) {
+            Log.e(TAG, "Exception generating payload:" + exception.toString());
+            PluginResult result = new PluginResult(
+                    PluginResult.Status.ERROR,
+                    "Error occurred while creating onMessage payload: " + e.toString()
+            );
+            CallbackUtil.sendPluginResult(onMessageCallback, result);
+        }
+    }
+
     private class DatagramSocketSend implements Runnable {
-        String m_message;
-        String m_address;
-        int m_port;
-        int m_datagramSocketId;
-        DatagramSocket m_datagramSocket;
-
-        public DatagramSocketSend(
-                final String message,
-                final String address,
-                final int port,
-                final int datagramSocketId,
-                final DatagramSocket datagramSocket
-        ) {
-            this.m_message = message;
-            this.m_address = address;
-            this.m_port = port;
-            this.m_datagramSocketId = datagramSocketId;
-            this.m_datagramSocket = datagramSocket;
+        CallbackContext sendCallback;
+        String message;
+        String address;
+        int port;
+
+        public DatagramSocketSend(CallbackContext callbackContext, final String message,
+                final String address, final int port) {
```

Comment:
nitpick / wouldn't it be nice if these plugins had linting?

not sure what the common Java style is, but I know in Kotlin, this would be preferred:

```
public DatagramSocketSend(
    CallbackContext callbackContext,
    final String message,
    final String address,
    final int port
) {

Reply:
Java is typically what I put. List args until you get to max line length and then wrap.